docs/01-Headless/03-Directives.md

```diff
@@ -1,5 +1,244 @@
+
+```html
+<div id="content">
+	<div id="focus-element">
+		<button id="test">button 1</button>
+		<button id="test2">button 2</button>
+	</div>
+	<hr />
+	<input id="clickText" type="text" />
+	<hr />
+	<button onclick="document.getElementById('focus-element').remove()">remove dom element</button>
+</div>
+```
+
+To use the directive in vanilla TS you need the following steps:
+
+1. Create the directive on a DOM element. In this case, our elementwill be the `focus-element` container.
+
+```typescript
+const trackElement = document.getElementById('focus-element');
+const focusElementDirective = createMyDirective(trackElement, 'focus-element click');
+```
+
+By doing it, when you click on a `button` inside the `focus-element`, you will get a log into your console,
+
+```
+focus-element click
+HTMLButtonElement{attributes: {…}, innerHTML: "button 1", nodeType: 1, tagName: "button"}
+```
+
+2. Observe context changes and call the `update` function accordingly. For example, we want to change the text to the value of the `clickText` input when it changes.
+
+```typescript
+const input = document.getElementById('clickText');
+input.addEventListener('change', (event) => {
+	focusElementDirective?.update(event.target.value);
+});
+```
+
+3. Clean up when the `focus-element` gets remove from the DOM. This can be done for example using a `MutationObserver` on the parent DOM element.
+
+```typescript
+const cleanup = (mutations: MutationRecord[]) => {
+	mutations.forEach((mutation) => {
+		for (let entry of mutation.removedNodes.entries()) {
+			if (entry.includes(trackElement)) {
+				focusElementDirective?.destroy();
+			}
+		}
+	});
+};
+const observer = new MutationObserver(cleanup);
+observer.observe(document.getElementById('content'), {childList: true});
+```
+
+### Angular usage
+
+By default, Angular provides a decorator that marks a class as an Angular directive. They are used to attach custom behavior to elements in the DOM and Directive classes can implement life-cycle hooks to influence their configuration behavior (source: Angular doc).
+
+Agnos provides an utility to use Agnos Directives in Angular, `[auUse]`.
+`[auUse]` directive takes as input an Agnos `Directive`, and takes care of execution, update and destroy of it. This thanks to the life-cycle events provided by Angular (`OnChanges`, `OnDestroy`).
+Unfortunately, a directive cannot take additional properties, hence parameters are given through a second input, `[auUseParams]`, which type is the same as the second argument of the Agnos Directive.
+During the `OnChanges` hook, the directive will be executed for the first time (if the reference of the `Directive`) or the Directive `update` function (if provided) will be executed with the new `auUseParams` input.
+
+#### Example
+
+```typescript
+import {UseDirective} from '@agnos-ui/angular-headless';
+import {Component} from '@angular/core';
+import {FormsModule} from '@angular/forms';
+
+@Component({
```

Comment:
Maybe we should highlight that we can use the same directive agnostic code for all framework...
Extract the directive code above the three framework ?